Initial Classes

mecha opened the pull request.

mecha committed:
Updated dependencies

mecha committed:
Added `Session` class

mecha committed:
Added `SessionGenerator`

mecha committed:
Removed `Session`

mecha committed:
Updated dependencies

mecha committed:
Fixed `generate()` signature

mecha committed:
Session factory is now required

XedinUnknown commented on src/SessionGenerator.php:
```diff
@@ -0,0 +1,303 @@
+            $validator->validate($session);
+
+            return true;
+        } catch (ValidationFailedExceptionInterface $e) {
```
Why does a method starting with "validate" return something like an `isValid()` method would? Perhaps, this is because the [interface](https://github.com/RebelCode/session-generator-interface/blob/b0a1045a8169a7f74506229f4e088ff060d7d4b3/src/SessionGeneratorInterface.php#L24) does not permit throwing. In this case, maybe an exception interface needs to be put in, and the generator should be allowed to throw.

mecha replied:
Given that the validation is used to filter which sessions are kept and which are discarded, having the generator throw does not make sense, in my opinion.

The validation _inside_ the generator should allow validation to fail so it can determine that a session is not meant to be yielded. It does this by catching `ValidationFailedExceptionInterface`.

You are right on the name; something like `isSessionValid()` would make more sense.

XedinUnknown replied:
I suggest that you throw as usual here, and catch in i.e. `_generate()`.

XedinUnknown replied:
@mecha, not sure if this is addressed yet.

mecha replied:
@XedinUnknown Yes it has been. All of your reviews on this PR have been outdated since 25th August 2017 😆

XedinUnknown commented on src/SessionGenerator.php:
```diff
@@ -0,0 +1,303 @@
+     * @param array $startTimes A temporary array for storing the starting times that have already been processed, to
+     *                          avoid repeating specific passes.
+     */
+    protected function _generate($rangeStart, $rangeEnd, array &$results = [], array &$startTimes = [])
```
Looks like something that belongs in an abstract class.

XedinUnknown replied:
@mecha, also not addressed, it seems.

mecha replied:
You are commenting on code that doesn't exist anymore bro! 😆 This is outdated.

XedinUnknown commented on src/SessionGenerator.php:
```diff
@@ -0,0 +1,303 @@
+     *
+     * @return int[] A list of session lengths.
+     */
+    public static function calculateLengths($minLength, $maxLength, $stepLength)
```
Looks like something that belongs in an abstract class.

XedinUnknown commented on src/SessionGenerator.php:
```diff
@@ -0,0 +1,303 @@
+        $this->_setSessionLengths($lengths)
+             ->_setPaddingTime($padding)
+             ->_setValidator($validator)
+             ->_setSessionFactory($sessionFactory);
```
🥇

mecha committed:
Added `SessionLengthsAwareTrait`

mecha committed:
Added `PaddingTimeAwareTrait`

mecha committed:
Added `CalculateSessionLengthsCapableTrait`

mecha committed:
Added `SessionFactoryAwareTrait`

mecha committed:
Added `ValidatorAwareTrait`

mecha committed:
Fixed code standards

mecha committed:
Updated dependencies

mecha committed:
Fixed code standards in `SessionGenerator` test

mecha committed:
Updated dev dependencies

mecha committed:
Added `AbstractSessionGenerator`

mecha committed:
Fixed `session-generator-interface` dependency

mecha committed:
Added `AbstractSessionGeneratorTest`

mecha committed:
Fxed `SessionGenerator` to comply with ...

... the latest changes to `AbstractSessionGenerator`

mecha committed:
Updated dependencies

mecha committed:
Fixed test for `AbstractSessionGenerator`

mecha committed:
Updated dependencies

- Allowing more versions of `dhii/validation-interface`

mecha committed:
Installed `dhii/factory-base`

mecha committed:
Added `SessionGeneratorFactory`

mecha committed:
Updated dependency version constraints

XedinUnknown committed:
Added test folders to `autoload-dev`

XedinUnknown committed:
`develop` now aliased to `0.1.x-dev`

XedinUnknown committed:
Increased dep stability; all deps updated

XedinUnknown committed:
Now building also on PHP 7.2

XedinUnknown committed:
Added changelog entry

XedinUnknown committed:
Brought readme up to standard

XedinUnknown committed:
Added PHPS project configs

XedinUnknown committed:
Increased dep stability; all deps updated

XedinUnknown committed:
Added test framework config to PHPS

XedinUnknown committed:
Fixed `SessionGeneratorFactory`

Was lacking methods, which are now implemented via traits.

XedinUnknown committed:
Fixed test of `SessionGeneratorFactory`

Was expecting `CouldNotMakeExceptionInterface`, but got a more generic
`FactoryExceptionInterface`. Not sure exactly which is right. A more
precise definition of `CouldNotMakeExceptionInterface` is needed.
